libcaf_core/src/binary_serializer.cpp

```diff
@@ -0,0 +1,186 @@
+ *                     | |     / _ \ | |_       Actor                         *
+ *                     | |___ / ___ \|  _|      Framework                     *
+ *                      \____/_/   \_|_|                                      *
+ *                                                                            *
+ * Copyright 2011-2018 Dominik Charousset                                     *
+ *                                                                            *
+ * Distributed under the terms and conditions of the BSD 3-Clause License or  *
+ * (at your option) under the terms and conditions of the Boost Software      *
+ * License 1.0. See accompanying files LICENSE and LICENSE_ALTERNATIVE.       *
+ *                                                                            *
+ * If you did not receive a copy of the license files, see                    *
+ * http://opensource.org/licenses/BSD-3-Clause and                            *
+ * http://www.boost.org/LICENSE_1_0.txt.                                      *
+ ******************************************************************************/
+
+#include "caf/binary_serializer.hpp"
+
+#include <iomanip>
+#include <sstream>
+
+#include "caf/detail/ieee_754.hpp"
+#include "caf/detail/network_order.hpp"
+
+namespace caf {
+
+namespace {
+
+template <class T>
+error apply_int(binary_serializer& bs, T x) {
+  auto y = detail::to_network_order(x);
+  return bs.apply_raw(sizeof(T), &y);
+}
+
+} // namespace <anonmyous>
+
+binary_serializer::binary_serializer(actor_system& sys, buffer& buf)
+  : super(sys), buf_(buf), write_pos_(buf.end()) {
+  // nop
+}
+
+binary_serializer::binary_serializer(execution_unit* ctx, buffer& buf)
+  : super(ctx), buf_(buf), write_pos_(buf.end()) {
+  // nop
+}
+
+void binary_serializer::seek(size_t offset) {
+  write_pos_ = buf_.begin() + offset;
+}
+
+void binary_serializer::skip(size_t num_bytes) {
+  auto last = buf_.end();
+  auto remaining = static_cast<size_t>(std::distance(write_pos_, last));
+  if (remaining >= num_bytes) {
+    write_pos_ += num_bytes;
+  } else {
+    buf_.insert(last, num_bytes - remaining, 0);
+    write_pos_ = buf_.end();
+  }
+}
+
+error binary_serializer::begin_object(uint16_t& nr, std::string& name) {
+  if (nr != 0)
+    return apply(nr);
+  if (auto err = apply(nr))
+    return err;
+  return apply(name);
+}
+
+error binary_serializer::end_object() {
+  return none;
+}
+
+error binary_serializer::begin_sequence(size_t& list_size) {
+  auto s = static_cast<uint32_t>(list_size);
+  return apply(s);
+}
+
+error binary_serializer::end_sequence() {
+  return none;
+}
+
+error binary_serializer::apply_raw(size_t num_bytes, void* data) {
+  auto ptr = reinterpret_cast<char*>(data);
+  auto last = buf_.end();
+  if (write_pos_ == last) {
+    buf_.insert(last, ptr, ptr + num_bytes);
+    write_pos_ = buf_.end();
+  } else if (write_pos_ + num_bytes < last) {
+    std::copy(ptr, ptr + num_bytes, write_pos_);
```

Comment:
I remember discussions with @initconf a few years ago where he found drastic performance differences between `std::copy` and `std::memcpy`, even though this shouldn't matter in a good implementation of the standard library. To avoid unexpected performance degradations on this critical path, I'd use `std::memcpy` instead.